blockchain/chainquery.go

```diff
@@ -132,3 +132,25 @@ func (b *BlockChain) ChainTips() []ChainTipInfo {
 	}
 	return results
 }
+
+// BestHeader returns the header with the most cumulative work that is NOT
+// known to be invalid.
+func (b *BlockChain) BestHeader() (chainhash.Hash, int64) {
+	b.index.RLock()
+	header := b.index.bestHeader
+	b.index.RUnlock()
+	return header.hash, header.height
+}
+
+// BestInvalidHeader returns the header with the most cumulative work that is
+// known to be invalid.  It will be a hash of all zeroes if there is no such
+// header.
+func (b *BlockChain) BestInvalidHeader() chainhash.Hash {
```

Comment:
What is the primary reason for tracking and exposing the best invalid header?

Reply:
For this PR, the main use is in the tests to help ensure all of the logic is working as expected in terms of invalidation.

However, in addition to helping ensure correctness, the main intention is that it will be useful once we transition to solely using headers for network synchronization.  At that point it can be used easily detect when the network has forked away via a hard fork vote and related consensus change which in turn will provide the ability to behave nicely in that circumstances as well as provide more understandable warnings and error messages that they need to upgrade.